bf: ZENKO-860, S3C-1820 maximize queue concurrency usage

jonathan-gramain opened the pull request:
Note for previous reviewers of this PR: the first commit is mostly the previous state of this PR that fixes the queue being blocked while one entry is slow to complete, except:

- removed a `setTimeout()` in case the queue is full, which was useless

- added a check with `isPaused()` to avoid calling `offsetsStore()` on a paused consumer, which causes an exception to be thrown because the consumer is not subscribed to a topic.

**bf: ZENKO-860 replace replication batching by continuous task flow**

Keep consuming kafka messages as soon as new workers become available
in the queue, rather than relying on the 'drain' callback from
async.queue.

This prevents being stuck from consuming new messages in case one of
the tasks takes a long time to finish (e.g. if retrying on a failing
cloud).

Change the way we manage Kafka consumer offsets: instead of disabling
auto-commit by default, always enable it and disable automatic offset
store of consumed messages. This allows to set the offset based on the
processed messages rather than the consumed messages, relying on a
helper class OffsetLedger to maintain knowledge of the highest offset
that can be committed, i.e. that is a higher-bound of already
processed message offsets.


**ft: ZENKO-860 improve replication task pool efficiency**

By reducing drastically the time it takes for a replication task to
complete, for the same concurrency number, a lot more tasks can be
executed, leading to more throughput and quicker completion of
replication queue tasks especially for small objects, while using
cloud resources more consistently and efficiently with the same bound
on the number of open connections.

To achieve this reduction in execution time, tasks do not wait anymore
for a delivery report to be returned from Kafka producers (for metrics
and backbeat-replication-status topics) before finishing. Most of the
wait time was spent waiting for the delivery reports, which are only
issued once Kafka commits changes in batch (every few seconds), hence
the tasks in progress were mostly waiting for those delivery reports
instead of transferring bytes of data between locations.

Instead, after the data replication is completed (success or failure),
and the task completes, we keep track of the delivery report pending
for the message produced to backbeat-replication-status topic. Only
when this status comes back (success or failure), we acknowledge that
this message has been consumed and processed, to allow eventually
committing the consumer group offset forward.

Delivery reports for messages produced to metrics topic are ignored,
because it does not seem to bring any consistency benefit to wait for
them (but we could instead implement a dedupe cache to get rid of
duplicates).

jonathan-gramain committed:
bf: ZENKO-860 replace replication batching by continuous task flow

Keep consuming kafka messages as soon as new workers become available
in the queue, rather than relying on the 'drain' callback from
async.queue.

This prevents being stuck from consuming new messages in case one of
the tasks takes a long time to finish (e.g. if retrying on a failing
cloud).

Change the way we manage Kafka consumer offsets: instead of disabling
auto-commit by default, always enable it and disable automatic offset
store of consumed messages. This allows to set the offset based on the
processed messages rather than the consumed messages, relying on a
helper class OffsetLedger to maintain knowledge of the highest offset
that can be committed, i.e. that is a higher-bound of already
processed message offsets.

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @vrancurel

* @rahulreddy

* @bennettbuchanan

* @jonathan-gramain

rahulreddy commented on lib/BackbeatConsumer.js:
```diff
@@ -224,18 +226,14 @@ class BackbeatConsumer extends EventEmitter {
             if (!this._autoCommit) {
                 this._consumer.commit();
             }
-            this.emit('consumed', this._messagesConsumed);
```
If autocommit is not enabled (which is the case for crr) and now that we are switching to flowing mode, when there are a lot of messages, the queue would never be empty (correct?) So the manual commit would probably happen after n messages and n could be 10/100/200?

jonathan-gramain replied:
We are actually not switching to flowing mode, we are still running in non-flowing mode by consuming a specific amount of messages (https://github.com/scality/backbeat/pull/399/files#diff-e2120238e136b010d97978588107a8cbR278).

The `'consumed'` event is right now only used in tests AFAICS, and I moved it from this place to https://github.com/scality/backbeat/pull/399/files#diff-e2120238e136b010d97978588107a8cbR278, where each ending task has an opportunity to emit this event.

jonathan-gramain replied:
I realize this deserves a more specific answer.

Even though we're still in non-flowing mode, you're still right in saying the queue may never get empty, this until we stop adding new messages to the queue at least. This behavior has the same kind of validity that it used to, but it may have become less useful since there's no guarantee that it will be called timely.

To overcome this, the caller either has to use auto-commit, or we should add a `commit()` wrapper that the user is responsible for calling when he wishes to commit the current offset. For our use-case it seems reasonable to use auto-commit for replication. It may be also worth enabling it by default, like the kafka client does.

jonathan-gramain replied:
This has been solved by enabling auto-commit on the kafka consumer.

rahulreddy commented on lib/BackbeatConsumer.js:
```diff
@@ -265,8 +263,20 @@ class BackbeatConsumer extends EventEmitter {
+        }
+        this._nConsumePendingRequests += nNewConsumeRequests;
+        return this._consumer.consume(nNewConsumeRequests, (err, entries) => {
+            this._nConsumePendingRequests -= nNewConsumeRequests;
```
`this._nConsumePendingRequests -= nNewConsumeRequests;` this can be decremented only if there is no error fetching the entries correct?

jonathan-gramain replied:
Actually I want to decrement it unconditionally, because this is the bookkeeper for pending *requests*, so even if the requests fail they are not pending anymore, and we don't want the counter to be off thereafter.

rahulreddy reviewed (changes requested):
Approved accidentally

JianqinWang commented on tests/functional/BackbeatConsumer.js:
```diff
@@ -43,37 +39,43 @@ describe('BackbeatConsumer', () => {
+            bootstrap: true,
         });
+        async.parallel([
+            done => producer.on('ready', done),
```
I assume that having the outer function callback named `done` does not conflict with the `async` callbacks also named `done` - it was still confusing to read at a first glance xD.

jonathan-gramain replied:
it does not conflict, though may be confusing indeed

philipyoo reviewed (dismissed):
lgtm 👍

philipyoo commented:
are tests failing because requires more time than current set timeout to complete?

jonathan-gramain commented:
The latest changes should address remaining concerns (about the offset commit especially).

jonathan-gramain commented:
Postponed to Zenko 1.1

jonathan-gramain committed:
ft: ZENKO-860 improve replication task pool efficiency

By reducing drastically the time it takes for a replication task to
complete, for the same concurrency number, a lot more tasks can be
executed, leading to more throughput and quicker completion of
replication queue tasks especially for small objects, while using
cloud resources more consistently and efficiently with the same bound
on the number of open connections.

To achieve this reduction in execution time, tasks do not wait anymore
for a delivery report to be returned from Kafka producers (for metrics
and backbeat-replication-status topics) before finishing. Most of the
wait time was spent waiting for the delivery reports, which are only
issued once Kafka commits changes in batch (every few seconds), hence
the tasks in progress were mostly waiting for those delivery reports
instead of transferring bytes of data between locations.

Instead, after the data replication is completed (success or failure),
and the task completes, we keep track of the delivery report pending
for the message produced to backbeat-replication-status topic. Only
when this status comes back (success or failure), we acknowledge that
this message has been consumed and processed, to allow eventually
committing the consumer group offset forward.

Delivery reports for messages produced to metrics topic are ignored,
because it does not seem to bring any consistency benefit to wait for
them (but we could instead implement a dedupe cache to get rid of
duplicates).

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.0/bugfix/ZENKO-860-backoff-block` with contents from `bugfix/ZENKO-860-backoff-block`
and `development/8.0`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.0/bugfix/ZENKO-860-backoff-block origin/development/8.0
 $ git merge origin/bugfix/ZENKO-860-backoff-block
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.0/bugfix/ZENKO-860-backoff-block
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-860-backoff-block` into
`development/7.4`
* [w/8.0/bugfix/ZENKO-860-backoff-block](https://github.com/scality/backbeat/compare/development/8.0...w/8.0/bugfix/ZENKO-860-backoff-block) will be merged into `development/8.0` (pull request #528)
* [w/8.1/bugfix/ZENKO-860-backoff-block](https://github.com/scality/backbeat/compare/development/8.1...w/8.1/bugfix/ZENKO-860-backoff-block) will be merged into `development/8.1` (pull request #529)





*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @vrancurel

* @jonathan-gramain

* @rahulreddy

* @bennettbuchanan

philipyoo commented on extensions/replication/tasks/ReplicateObject.js:
```diff
@@ -132,33 +132,39 @@ class ReplicateObject extends BackbeatTask {
+                this.consumer.onEntryCommittable(kafkaEntry);
+            }
+        });
+        log.end().info('replication status published', {
```
maybe only log this if an error didn't occur above https://github.com/scality/backbeat/pull/399/commits/63b260d0b2827d15501f984815fd0ace8fae6b76#diff-e5fe2d1a77d806ec6632afc36abf102fR145

jonathan-gramain replied:
This is logged in the synchronous context, so at this point we can only assert that we have published but did not get the feedback yet.

philipyoo reviewed (approved):
lgtm 👍

jonathan-gramain commented:
Manual run of backbeat E2E tests passes. (using custom branch test/backportCRRRetryZenkoE2ETests and my two lab deployments, I gave up for now trying to run them as part of eve automation).

It has also been tested with large workloads (100000+ objects) without issues.

rahulreddy commented on extensions/replication/tasks/ReplicateObject.js:
```diff
@@ -132,33 +132,39 @@ class ReplicateObject extends BackbeatTask {
-            return cb();
+            // Commit whether there was an error or not to allow
+            // progress of the consumer, as best effort measure when
+            // there are errors. We can count on the sweeper to retry
```
When you say `sweeper`, do you mean the background task that publishes stalled entries? If yes, that would mean that we have to enable it by default in Zenko and install something comparable in S3C as well.

jonathan-gramain replied:
Indeed we should have a sweeper for those cases. Note that the behavior here has not changed with this PR: before we were simply committing when 'drain' even was issued, whether there were errors processing entries or not. So I think we should consider this separately from this PR, as it's not directly related.

rahulreddy commented on lib/BackbeatConsumer.js:
```diff
@@ -191,29 +186,47 @@ class BackbeatConsumer extends EventEmitter {
             if (err || entries.length === 0) {
-                // retry later to fetch new messages in case no one is
-                // available yet in the message queue
+                // no message is available yet, retry later
```
Let's add a debug log here for visibility

rahulreddy replied:
It's worth logging debug log here.

rahulreddy commented on lib/BackbeatConsumer.js:
```diff
@@ -65,28 +68,31 @@ class BackbeatConsumer extends EventEmitter {
         this._processingQueue = null;
         this._messagesConsumed = 0;
-
+        this._nConsumePendingRequests = 0;
```
Can you please comment what this property represents?

rahulreddy commented on lib/BackbeatConsumer.js:
```diff
@@ -191,29 +186,47 @@ class BackbeatConsumer extends EventEmitter {
-        this._consumer.consume(this._concurrency, (err, entries) => {
+        // needs explicit polling to get new messages
+
+        const nNewConsumeRequests =
```
Can you make this a method that is more descriptive? This represents the queue status and we use it to decide if more items can be pushed to the queue.

rahulreddy commented on lib/OffsetLedger.js:
```diff
@@ -0,0 +1,143 @@
+            return partitionOffsets.processing[0];
+        }
+        if (partitionOffsets.latestConsumed !== null) {
+            return partitionOffsets.latestConsumed + 1;
```
What if this is somehow a `NaN` because `latestConsumed` is `undefined` 😄? Does `noderdkafaka` throw an exception in this case?

jonathan-gramain replied:
Not sure it can happen, but I can make it bullet-proof by checking that it's an integer instead.

rahulreddy commented on lib/OffsetLedger.js:
```diff
@@ -0,0 +1,143 @@
+            // offsets currently being processed, as they will be
+            // reprocessed
+            partitionOffsets.processing =
+                partitionOffsets.processing.filter(pOff => pOff < offset);
```
When you say higher offset, isn't the conditional `pOff > offset`?

jonathan-gramain replied:
`filter` will keep the lower offsets only, so getting rid of the higher ones.

rahulreddy commented on tests/functional/lib/BackbeatConsumer.js:
```diff
@@ -0,0 +1,221 @@
+        async.series([
+            next => {
+                setTimeout(() => producer.send(boatloadOfMessages, err => {
+                    assert.ifError(err);
```
This is asynchronous and the code following this i synchronous, I think it deserves its own step.

jonathan-gramain replied:
The synchronous code executes immediately without waiting for`producer.send` to complete, this is done on purpose so that we produce messages only after we are ready to consume them. If you have another preference for the code style let me know.

rahulreddy commented on tests/functional/lib/BackbeatConsumer.js:
```diff
@@ -0,0 +1,221 @@
+                }, 2000);
+            }
+        });
+    }).timeout(30000);
```
For timeouts, let's set them globally at once. It can be done in package.json when invoked by command or mocha.opts.

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.0/bugfix/ZENKO-860-backoff-block` with contents from `bugfix/ZENKO-860-backoff-block`
and `development/8.0`.

Please resolve the conflict on **the integration branch** (`w/8.0/bugfix/ZENKO-860-backoff-block`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.0/bugfix/ZENKO-860-backoff-block
 $ git pull  # or "git reset --hard origin/w/8.0/bugfix/ZENKO-860-backoff-block"
 $ git merge origin/development/8.0
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/bugfix/ZENKO-860-backoff-block
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.0/bugfix/ZENKO-860-backoff-block
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @vrancurel

* @jonathan-gramain

* @rahulreddy

* @bennettbuchanan

bert-e commented:
# History mismatch



Merge commit #280c1bd61e74a5640ec737928bce0867e790c4cd on the integration branch
`w/8.0/bugfix/ZENKO-860-backoff-block` is merging a branch which is neither the current
branch `bugfix/ZENKO-860-backoff-block` nor the development branch
`development/8.0`.

It is likely due to a rebase of the branch `bugfix/ZENKO-860-backoff-block` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.0/bugfix/ZENKO-860-backoff-block` with contents from `bugfix/ZENKO-860-backoff-block`
and `development/8.0`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.0/bugfix/ZENKO-860-backoff-block origin/development/8.0
 $ git merge origin/bugfix/ZENKO-860-backoff-block
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.0/bugfix/ZENKO-860-backoff-block
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-860-backoff-block` into
`development/7.4`
* [w/8.0/bugfix/ZENKO-860-backoff-block](https://github.com/scality/backbeat/compare/development/8.0...w/8.0/bugfix/ZENKO-860-backoff-block) will be merged into `development/8.0` (pull request #539)
* [w/8.1/bugfix/ZENKO-860-backoff-block](https://github.com/scality/backbeat/compare/development/8.1...w/8.1/bugfix/ZENKO-860-backoff-block) will be merged into `development/8.1` (pull request #540)





*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @vrancurel

* @jonathan-gramain

* @rahulreddy

* @bennettbuchanan

jonathan-gramain commented:
Zenko tests ran and passed here, with a custom backbeat image: https://eve.devsca.com/github/scality/zenko/#/builders/2/builds/1046

jonathan-gramain commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.4`

* :heavy_check_mark: `development/8.0`

* :heavy_check_mark: `development/8.1`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.4`

* :heavy_check_mark: `development/8.0`

* :heavy_check_mark: `development/8.1`




Please check the status of the associated issue ZENKO-860.

Goodbye jonathan-gramain.
